### PR TITLE
Request based workers

### PR DESCRIPTION
**Pull request by the-allanc**
❓ **What kind of change does this PR introduce?**
  - [X] 🐞 bug fix
  - [x] 🐣 feature
  - [ ] 📋 docs update
  - [ ] 📋 tests/coverage improvement
  - [ ] 📋 refactoring
  - [ ] 💥 other

📋 **What is the related issue number (starting with `#`)**

#91 

❓ **What is the current behavior?**

Worker threads are allocated a connection object, and that thread is occupied by that connection is closed or times out.

This can lead to the server having no spare threads to handle new incoming connections - even if there are no further requests coming over the used connections.

❓ **What is the new behavior (if this is a feature change)?**

We maintain a connection pool of open connections. Cheroot uses `select` to determine if any of the open connections or the server socket has any incoming data - we then either use an existing connection or accept a new connection, and pass that to the worker thread.

Once the worker thread has processed a request, it will pass the connection back to the connection pool (if it decides that the connection should not be closed). This allow threads to only be occupied by connections that are ready with a request.

📋 **Other information**:

The connection pool uses two settings - a new one to determine how many keep-alive connections to keep open (this only applies to connections which are not currently being handled by a thread).

It also uses the server socket timeout setting to determine when to close an idle keep-alive connection.

📋 **Checklist**:

  - [x] I think the code is well written
  - [x] I wrote [good commit messages][1]
  - [x] I have [squashed related commits together][2] after the changes have been approved
  - [x] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [x] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/199)

### CODE REVIEW

File: cheroot/connections.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
+"""Utilities to manage open connections."""
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+import io
+import select
+import six
```
Comment (webknjaz) [inline cheroot/connections.py]: this is a third-party lib, it should go into a separate section

File: cheroot/connections.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+import socket
+import time
+
+from . import errors
+from .makefile import MakeFile
+
+from collections import OrderedDict
```
Comment (webknjaz) [inline cheroot/connections.py]: this is an import from stdlib, it should go into the top group of imports

File: cheroot/connections.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+
+try:
```
Comment (webknjaz) [inline cheroot/connections.py]: I feel like this whole try/except cascade should probably go under `_compat` module.

**Comment by webknjaz** [comment]
@the-allanc have you seen https://docs.python.org/3/library/socketserver.html ? Is it applicable to Cheroot project somehow?

File: cheroot/connections.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+        This happens if there are either too many open connections, or if the
+        connections have been timed out.
+
+        This should be called periodically.
+        """
+        if not self.connections:
+            return
+
+        # Check for too many open connections.
```
Comment (webknjaz) [inline cheroot/connections.py]: It probably makes sense to check for expired connections first. This will allow keeping more of non-expired connections alive.

File: cheroot/connections.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        ka_limit = self.server.keep_alive_conn_limit
+        if ka_limit is not None:
+            [self._close(conn[0]) for conn in conns[:-ka_limit]]
+            conns = conns[-ka_limit:]
+
+        # Check for old connections.
+        now = time.time()
+        for (conn, (ctime, _)) in conns:
+            # Oldest connection out of the currently available ones.
+            if (ctime + self.server.timeout) < now:
```
Comment (webknjaz) [inline cheroot/connections.py]: It probably makes sense to add `expired` property to the connection itself which would turn this check into `if conn.expired:`

File: cheroot/connections.py (hunk 4 above)
Comment (webknjaz) [reply cheroot/connections.py]: Oh, after looking closer: expired connections would go first anyway...

File: cheroot/connections.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        """
+        if not self.connections:
+            return
+
+        # Check for too many open connections.
+        conns = list(self.connections.items())
+        ka_limit = self.server.keep_alive_conn_limit
+        if ka_limit is not None:
+            [self._close(conn[0]) for conn in conns[:-ka_limit]]
+            conns = conns[-ka_limit:]
```
Comment (webknjaz) [inline cheroot/connections.py]: Maybe
```suggestion
            del conns[:-ka_limit]
```
?

File: cheroot/connections.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+        # Check for old connections.
+        now = time.time()
+        for (conn, (ctime, _)) in conns:
+            # Oldest connection out of the currently available ones.
+            if (ctime + self.server.timeout) < now:
+                self._close(conn)
```
Comment (webknjaz) [inline cheroot/connections.py]: You close connections but don't delete them from `self.connections`, do you? Wouldn't this cause a memory leak?

Comment (webknjaz) [reply cheroot/connections.py]: Oh, you just remove them inside of that call...

File: cheroot/connections.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        self.connections = OrderedDict()
+
+    def put(self, conn):
+        """Put idle connection into the ConnectionManager to be managed.
+
+        Args:
+            conn (cheroot.server.HTTPConnection): HTTP connection
+                to be managed.
+        """
+        self.connections[conn] = time.time(), conn.rfile.has_data()
```
Comment (webknjaz) [inline cheroot/connections.py]: How about having a [Circular buffer](https://en.wikipedia.org/wiki/Circular_buffer) for this?

Comment (webknjaz) [reply cheroot/connections.py]: Maybe inherit `deque`? https://stackoverflow.com/a/4151368/595220

File: cheroot/connections.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        # Check for too many open connections.
+        conns = list(self.connections.items())
+        ka_limit = self.server.keep_alive_conn_limit
+        if ka_limit is not None:
+            [self._close(conn[0]) for conn in conns[:-ka_limit]]
+            conns = conns[-ka_limit:]
+
+        # Check for old connections.
+        now = time.time()
+        for (conn, (ctime, _)) in conns:
```
Comment (webknjaz) [inline cheroot/connections.py]: This looks like a perfect use case for [`itertools.takewhile()`](https://docs.python.org/3/library/itertools.html#itertools.takewhile)

File: cheroot/connections.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+        Any connection returned by this method will need to be `put`
+        back if it should be examined again for another request.
+
+        Args:
+            server_socket: ServerSocket instance to listen to for new
+            connections.
+        Returns:
+            cheroot.server.HTTPConnection instance, or None.
+
```
Comment (webknjaz) [inline cheroot/connections.py]:
```suggestion
```

File: cheroot/connections.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        A connection returned by this method should be ready for a worker
+        to handle it. If there are no connections ready, None will be
+        returned.
+
+        Any connection returned by this method will need to be `put`
+        back if it should be examined again for another request.
+
+        Args:
+            server_socket: ServerSocket instance to listen to for new
```
Comment (webknjaz) [inline cheroot/connections.py]:
```suggestion
            server_socket (socket.socket): ServerSocket instance to listen to for new
```

File: cheroot/connections.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+            socket_dict[conn.socket.fileno()] = conn
+        else:
+            # No ready connection.
+            conn = None
+
+        # Will require a select call.
+        if not conn:
+            ss_fileno = server_socket.fileno()
+            socket_dict[ss_fileno] = server_socket
+            rlist, _, _ = select.select(list(socket_dict), [], [], 0.1)
```
Comment (webknjaz) [inline cheroot/connections.py]: Is this run from the main thread?

File: cheroot/connections.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+        for conn, (tstamp, has_data) in self.connections.items():
+            if has_data:
+                break
+            socket_dict[conn.socket.fileno()] = conn
+        else:
+            # No ready connection.
+            conn = None
+
+        # Will require a select call.
+        if not conn:
```
Comment (webknjaz) [inline cheroot/connections.py]: You could revert the guard expression to be `if conn:` and return early. This would improve readability by reducing the indentation of the whole code block following this line.

File: cheroot/connections.py [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+                    s.settimeout(self.server.timeout)
+
+            conn = self.server.ConnectionClass(self.server, s, mf)
+
+            if not isinstance(self.server.bind_addr, six.string_types):
+                # optional values
+                # Until we do DNS lookups, omit REMOTE_HOST
+                if addr is None:  # sometimes this can happen
+                    # figure out if AF_INET or AF_INET6.
+                    if len(s.getsockname()) == 2:
```
Comment (webknjaz) [inline cheroot/connections.py]: How about:
```suggestion
                    addr = '::', 0
                    if len(s.getsockname()) == 2:
                        addr[0] = '0.0.0.0'
```

File: cheroot/server.py [hunk 15]
```diff
@@ -1978,99 +1947,18 @@ def resolve_real_bind_addr(socket_):
...
+        if not self.ready:
+            return
 
+        conn = self.connections.get_conn(self.socket)
+        if conn:
             try:
                 self.requests.put(conn)
             except queue.Full:
                 # Just drop the conn. TODO: write 503 back?
                 conn.close()
```
Comment (webknjaz) [inline cheroot/server.py]: Should this happen through the manager?

File: cheroot/connections.py (hunk 12 above)
Comment (the-allanc) [reply cheroot/connections.py]: Yes.

File: cheroot/connections.py (hunk 14 above)
Comment (the-allanc) [reply cheroot/connections.py]: I only moved this existing code from `server.py` to `connections.py`, so I don't want to get involved with making changes to that code (at least, not in this pull request).

File: cheroot/connections.py (hunk 3 above)
Comment (the-allanc) [reply cheroot/connections.py]: No comment on that - I just moved the code from `server.py` to `connections.py`.

File: cheroot/connections.py (hunk 10 above)
Comment (the-allanc) [reply cheroot/connections.py]: And I did initially have no whitespace there, but then [this CI job](https://travis-ci.org/cherrypy/cheroot/jobs/529425494) failed because of it.

Is there some tool that we can use to deal with auto-format these things prior to commit?

File: cheroot/server.py (hunk 15 above)
Comment (the-allanc) [reply cheroot/server.py]: The manager is responsible for providing a connection, but the server is then responsible for getting a worker thread to work on it. So no.

File: cheroot/connections.py (hunk 13 above)
Comment (the-allanc) [reply cheroot/connections.py]: I've changed this as suggested.

File: cheroot/connections.py (hunk 11 above)
Comment (the-allanc) [reply cheroot/connections.py]: Fixed.

File: cheroot/connections.py (hunk 9 above)
Comment (the-allanc) [reply cheroot/connections.py]: I've changed the behaviour of expire, so we don't loop like this any more.

File: cheroot/connections.py (hunk 8 above)
Comment (the-allanc) [reply cheroot/connections.py]: I did have this in earlier implementations. The problem is that we want a queue when checking if connections are still valid, but we want a hash-based lookup when we want to remove the connection object from the manager.

File: cheroot/connections.py (hunk 6 above)
Comment (the-allanc) [reply cheroot/connections.py]: This code doesn't exist any more.

File: cheroot/connections.py (hunk 5 above)
Comment (the-allanc) [reply cheroot/connections.py]: I've done something like that in the most recent set of changes.

**Comment by the-allanc** [comment]
> @the-allanc have you seen https://docs.python.org/3/library/socketserver.html ? Is it applicable to Cheroot project somehow?

I use the term "server socket" as a relic of my days programming in Java -[ServerSocket is a class](https://docs.oracle.com/javase/7/docs/api/java/net/ServerSocket.html) which is used for listening for and accepting incoming connections.

If some other term is appropriate, then please change it. It's not related to the SocketServer class in Python.

File: cheroot/connections.py (hunk 2 above)
Comment (the-allanc) [reply cheroot/connections.py]: Done.

File: cheroot/connections.py (hunk 1 above)
Comment (the-allanc) [reply cheroot/connections.py]: Done.

**Comment by the-allanc** [comment]
I've changed the code a bit in a few ways:

1. We store properties on the `HTTPConnection` object to indicate if it has data ready on the socket, if it's closeable and when it can be expired/
2. We now only expire one connection at a time (the "oldest" one), and let the main Cheroot loop "tick" and to then process that connection.
3. WorkerThreads are now responsible for closing connections, rather than the main Cheroot thread (which is how it used to work previously).

File: cheroot/connections.py (hunk 10 above)
Comment (jaraco) [reply cheroot/connections.py]: Yes. I suggest using [pipx](https://pypi.org/project/pipx) to install [pre-commit](https://pypi.org/project/pre-commit), then `pre-commit install`. This will run pre-commit before committing code. Alternately, for a lighter touch, just run `tox -e pre-commit`, which will run pre-commit manually. Sometimes, pre-commit will automatically edit the files. Other times, it can only raise an error requiring manual intervention.

**Comment by webknjaz** [comment]
@the-allanc FTR I'm having an HTTP parser migration in my mind (https://github.com/cherrypy/cheroot/issues/201). How well will this patch play with that?

File: cheroot/connections.py (hunk 8 above)
Comment (webknjaz) [reply cheroot/connections.py]: @the-allanc It would be nice if you could add a comment about this decision to the source code as well.

**Comment by webknjaz** [comment]
@the-allanc have you seen my question @ https://github.com/cherrypy/cheroot/pull/199#issuecomment-490317311?

File: cheroot/connections.py (hunk 3 above)
Comment (webknjaz) [reply cheroot/connections.py]: Yes, but since you're already moving it plz put it under _compat and just import this func from there making this module cleaner.

Comment (webknjaz) [reply cheroot/connections.py]: I'm not asking you to change it but I think it doesn't belong here.

File: cheroot/connections.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+    def __init__(self, server):
+        """Initialize ConnectionManager object.
+
+        Args:
+            server (cheroot.server.HTTPServer): web server object
+                that uses this ConnectionManager instance.
+        """
+        self.server = server
+        self.connections = OrderedDict()  # using this as a set
```
Comment (webknjaz) [inline cheroot/connections.py]: If you use it as a set why not use set? If you're using it for more, put a comment here explaining why

File: cheroot/makefile.py [hunk 17]
```diff
@@ -235,6 +235,7 @@ def readline(self, size=-1):
                         break
                     buf.write(data)
                 return buf.getvalue()
+
```
Comment (webknjaz) [inline cheroot/makefile.py]: I'm pretty sure this change is unrelated.

**Comment by webknjaz** [comment]
I believe that @mar10 will be interested in testing this out as well.

**Comment by the-allanc** [comment]
> @the-allanc FTR I'm having an HTTP parser migration in my mind (#201). How well will this patch play with that?

Had a quick look at it, I think generally it should work. Some minor changes required, like the connection manager needing a different way to determine if there's data in the buffer.

> @the-allanc have you seen my question @ [#199 (comment)](https://github.com/cherrypy/cheroot/pull/199#issuecomment-490317311)?

[Yes.](https://github.com/cherrypy/cheroot/pull/199#issuecomment-491367392)

**Comment by the-allanc** [comment]
After testing with my previous changes, it seems that `OrderedDict` isn't thread-safe as I originally thought - I think this would explain some weird errors we were seeing. So after some thought, I've decided just to use a basic list to maintain the collection of connections.

@jaraco - do you think we can get another dev release to try out?

**Comment by jaraco** [comment]
I made [this new release](https://m.devpi.net/jaraco/dev/cheroot/6.5.6.dev21+ge81bd245). I don't understand why it's `dev21` whereas the previous release was `dev28`.

**Comment by jaraco** [comment]
Aha - because dev28 was part of a different prerelease for a different branch (9413ed9c).

**Comment by webknjaz** [comment]
> because dev28 was part of a different prerelease

It's also tricky with history rewrites: since it's a distance to the last tag in chain squashing with force push will have a similar effect...

**Comment by webknjaz** [comment]
@the-allanc why isn't socketserver module applicable for Cheroot? It's in stdlib and I think I saw Werkzeug using it. It seemed to me that if there's something as high-level in the stdlib it might not be worth it to have an own low-level networking implemented: does it have any downsides?

**Comment by webknjaz** [comment]
@the-allanc what's not thread-safe with OrderedDict? Weren't you only manipulating it from a single thread?

File: cheroot/connections.py [hunk 18]
```diff
@@ -133,7 +132,7 @@ def get_conn(self, server_socket):
 
         # We have a connection ready for use.
         if conn:
-            del self.connections[conn]
+            self.connections.remove(conn)
```
Comment (webknjaz) [inline cheroot/connections.py]: Removal of an item from a list in Python is expensive in general. Especially, from its beginning: all following elements are shifted in memory. So OrderedDict was a more effective solution.

File: cheroot/connections.py [hunk 19]
```diff
@@ -154,12 +153,11 @@ def get_conn(self, server_socket):
                 try:
                     os.fstat(fno)
                 except OSError:
-                    # Socket is invalid, close the connection.
+                    # Socket is invalid, close the connection, insert at
+                    # the front.
+                    self.connections.remove(conn)
+                    self.connections.insert(0, conn)
```
Comment (webknjaz) [inline cheroot/connections.py]: Inserting things into the very beginning will shift the *entire* array in memory.

Comment (webknjaz) [reply cheroot/connections.py]: (https://stackoverflow.com/a/36529893/595220)

**Comment by jaraco** [comment]
I'm inclined to accept this even with a potentially performance-sensitive data structure like a list. The number of connections will typically be on the order of tens or hundreds, so seems unlikely to be performance sensitive. If the code is working in the production environment where the-allanc is piloting this code, I'd consider that sufficient to validate it as performance-safe in most production environments.... and we can optimize after.

I do think a [collections.deque](https://docs.python.org/3.7/library/collections.html#collections.deque) may be what @webknjaz has in mind for a more performant structure. If that's easy enough to use in place of a list without complicating the semantics, it's probably a good choice.

**Comment by webknjaz** [comment]
@jaraco even if I stop blocking this, the CIs (AppVeyor and Travis) should be made green first.

**Comment by webknjaz** [comment]
I looked again. It's acceptable now but still needs CIs to be green before merging.

**Comment by the-allanc** [comment]
> @the-allanc why isn't socketserver module applicable for Cheroot? It's in stdlib and I think I saw Werkzeug using it. It seemed to me that if there's something as high-level in the stdlib it might not be worth it to have an own low-level networking implemented: does it have any downsides?

Let me clarify - I explained why I used the term "server socket" in [this comment](https://github.com/cherrypy/cheroot/pull/199#issuecomment-491367392). It's not directly related to the `socketserver` module. I could have called it `incoming_connection_socket`, and we wouldn't be having this conversation.

Now if you want to use the `socketserver` module, that's your choice. It's been there since Python 1.5 (and possibly earlier), but CherryPy was developed after that and yet chose to ignore its existence. Maybe there were good reasons for that. But moving to that module is something that can be considered outside this merge request.

File: cheroot/connections.py (hunk 18 above)
Comment (the-allanc) [reply cheroot/connections.py]: > Weren't you only manipulating it from a single thread?
No. The main thread was cycling through open connections, but worker threads were also putting back non-closed connections.

File: cheroot/connections.py (hunk 19 above)
Comment (the-allanc) [reply cheroot/connections.py]: This isn't an action we would expect to be performing often.

**Comment by the-allanc** [comment]
>> I do think a [collections.deque](https://docs.python.org/3.7/library/collections.html#collections.deque) may be what @webknjaz has in mind for a more performant structure. If that's easy enough to use in place of a list without complicating the semantics, it's probably a good choice.

[From the documentation](https://docs.python.org/2/library/collections.html#collections.deque) on the page for `collections.deque`:
> Indexed access is O(1) at both ends but slows to O(n) in the middle. For fast random access, use lists instead.

We do both - adding to the end of the structure happens when reusable connections are put back, popping from the start of the structure happens when a connection needs to be expired, but connections are also removed from the middle of the structure itself.

I'm agnostic regarding what structure is used (`list` or `deque`), but I feel it's worth pointing out that it's by no means clear that a `deque` is better than a `list` for this code.

**Comment by webknjaz** [comment]
@the-allanc AFAIU `deque` is more efficient for removing items from the start of the collection because this operation doesn't require shifting each and every item in the memory.

**Comment by webknjaz** [comment]
As for the `socketserver`, I noticed it in the recent Werkzeug versions and was just curious. It's definitely unrelated to this PR :)

**Comment by mar10** [comment]
Hi, 
it seems that this PR is 95% complete and seems to bring major improvements to speed in some scenarios. Are there any estimations, when you can merge it?

**Comment by jaraco** [comment]
The main blocker here is figuring out why the tests are failing and getting them passing. If they're failing in the master, then we need to either fix those failures or mark them as xfail and file a separate ticket to fix those failures (with a preference being to fix them properly in order to maintain coverage). Then, merge the passing master into these proposed changes and assert stability.

If instead the test failures are due to the proposed changes, someone needs to investigate why and what needs to happen to the test suite to allow tests to pass without degrading coverage.

**Comment by webknjaz** [comment]
(triggered CIs)

**Comment by webknjaz** [comment]
pre-commit check, which has been failing earlier is no longer red.

**Comment by webknjaz** [comment]
It looks like `cheroot.test.test_ssl.test_http_over_https_error[::-builtin]` is persistently failing on master in AppVeyor under Python 2.7 only.

**Comment by webknjaz** [comment]
Alright, I kinda patched that..

**Comment by webknjaz** [comment]
Going to rebase this branch.

**Comment by webknjaz** [comment]
Looks like this PR makes some of TLS tests unstable under Windows: https://ci.appveyor.com/project/CherryPy/cheroot/build/job/bghdcuv66ri0o1pk

**Comment by webknjaz** [comment]
@the-allanc I've rebased this branch and solved the conflicts.

**Comment by h-vetinari** [comment]
Any update on this? :)

File: cheroot/server.py [hunk 20]
```diff
@@ -1255,30 +1260,26 @@ def __init__(self, server, sock, makefile=MakeFile):
...
-                    return
+            req = self.RequestHandlerClass(self.server, self)
+            req.parse_request()
+            if self.server.stats['Enabled']:
+                self.requests_seen += 1
+            if not req.ready:
+                # Something went wrong in the parsing (and the server has
+                # probably already made a simple_response). Return and
+                # let the conn close.
+                return
```
Comment (webknjaz) [inline cheroot/server.py]: I don't recall if I asked this before but it feels like this should return `False` explicitly. WDYT?

Comment (the-allanc) [reply cheroot/server.py]: I agree, I've made changes for it.

**Comment by jaraco** [comment]
I've tested this branch on my Windows box and the Python 2.7 errors happen on master same as on the branch, so I believe this branch is safe to merge.

**Comment by webknjaz** [comment]
Looks like this change introduced the following regressions:
* https://github.com/cherrypy/cheroot/issues/249
* https://github.com/cherrypy/cheroot/issues/263

**Comment by jaraco** [comment]
@the-allanc As I imagine you're aware, this change has caused some fairly serious regressions in the stability of cheroot (#249, #263). I suspect they'll be solved by the same investigation. I've spent some time in #263 writing a regression test that captures the failure and expected outcome. Will you have time to investigate these issues in the next few weeks? If not, I may take a stab at it, but if neither of us can make progress on it in fairly short order, I'm going to have to revert this feature.
